### DISCUSSION TITLE
Views Product Spec

### DISCUSSION THREAD

**Discussion by kgodey** [mathesar-foundation/mathesar#1015] [Maintainer discussions]
I've completed a large revision of the [Product Spec for Views](https://wiki.mathesar.org/en/product/specs/2022-01-views) and I'd like everyone to read and review it. The data explorer and Views are core features that will define how we do a lot of things in Mathesar and I'd like to catch problems with it early on.

I'll create a topic for each section below. Please feel free to provide any feedback you might have, including half-baked thoughts. I will provide specific prompts in each section in case that helps stimulate thoughts. You are also welcome to start your own topics if they don't fit into a specific page or if you have a lot to say.
- [Overview](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2277191)
- [Assumptions and Limitations](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2277197)
- [Feature Requirements](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2277202)
- [The Query Builder](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2277217)
- [Formulas](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2277225)
- [Views](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2277255)
- [Extra Information](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2277258)
- [2022-03-08 Spec Updates](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2319813)
- [2022-03-09 Spec Updates](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2325454)

## Previous discussion
- https://github.com/centerofci/mathesar-wiki/pull/22

**Comment by kgodey** [comment]
## Overview

["Views in Mathesar" wiki page](https://wiki.mathesar.org/en/product/specs/2022-01-views)

Prompts:
- Do the user goals of Views and Queries make sense?

**Reply by dmos62** [reply]
>The user goals for Views are:
> - To help users save commonly used queries

This makes me think of the number of views that we expect the user to have. In my experience with spreadsheets, I create a _lot_ of views, not because I commonly use each of them, but because other views depend on them. In other words, I create pipelines of views (or sheets as you'd call it in Google Sheets, for example). A good query engine would help keep the number of auxilliary views down (it can get obscenely long in Google Sheets), but you'd still need a view where there's a fork in the pipeline, for example.

I'm mentioning this, because I think it would be prudent for Mathesar to have good UX for working with a lot of views (and account for the use of auxialliary views). As I said, a good query engine will keep the number of views down, but if it has shortcoming a user will likely be able to compensate by creating more views.

**Reply by ghislaineguerin** [reply]
@dmos62 Are auxiliary views created as an in-between step to the final view? In this case, are you envisioning columns of the auxiliary views being selected in queries?

**Reply by mathemancer** [reply]
@ghislaineguerin I can't speak for Dom, but I'd definitely want that when using the product personally. It lets you create the view you eventually want iteratively.

As for the prompt, there's an assumption in the first paragraph of the page that users will generally be working with a well normalized database schema. I don't think that's necessarily true. Not only will many users not necessarily understand the value of normalizing their data (and so wouldn't do it), but some may actually have a good reason for a _de_normalized setup (e.g., if the main purpose of their DB is serving OLAP needs). While I didn't notice any big red flags in the more detailed points, we should be careful not to build that assumption into our goals and the designs that flow from them. It's just as valid to use views to process a big denormalized table as it is to get a denormalized view of a bunch of well normalized tables.

**Reply by dmos62** [reply]
@ghislaineguerin 

> Are auxiliary views created as an in-between step to the final view?

yes, the auxilliary views are in-between steps in the transformation towards the final view. The idea is to comfortably do the transformation from original tables to the final view in more than one step.

> Are you envisioning columns of the auxiliary views being selected in queries?

Yes, selected verbatim and also used to build/derive new columns. Like, I might have a view that cleans some data (e.g. removes currency symbols from number strings and converts them to the number type) and then a dependent view that adds those number columns together, or something along those lines.

To add to what @mathemancer said, I think it would be a normal use-case to use views to normalize some data and then to denormalize it again according to one's needs. That's been a common scenario for me; the "initial normalization" of obtained data is often not optimal. In such cases you end up with a lot of transformation steps, which, depending on the query engine, can end up as a lot of views.

**Reply by kgodey** [reply]
Auxiliary views are addressed in [this comment in the "Assumptions and Limitations"](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2288708) discussion.

> there's an assumption in the first paragraph of the page that users will generally be working with a well normalized database schema.

The main assumption is actually that we will rely on foreign keys (for now) to join data between tables. If the table is already denormalized, no foreign keys are necessary and everything should work fine.

We can support joining on natural keys in the future, but I wanted to keep the scope small for the alpha release.

> I might have a view that cleans some data (e.g. removes currency symbols from number strings and converts them to the number type) and then a dependent view that adds those number columns together, or something along those lines.

For now, you'd have to do that all in the same view. Perhaps we could have an option to hide columns in the alpha release to make it smoother (but easy to implement).

**Reply by pavish** [reply]
> Do the user goals of Views and Queries make sense?

They make good sense to me

**Reply by silentninja** [reply]
> Do the user goals of Views and Queries make sense?

Yes, they make sense

**Comment by kgodey** [comment]
## Assumptions and Limitations

[01. Assumptions and Limitations](https://wiki.mathesar.org/en/product/specs/2022-01-views/01-assumptions)

Prompts:
- Do these limitations and assumptions make sense? 
- Is there anything in this page where the wording is confusing or ambiguous?

**Reply by dmos62** [reply]
> We will not support using Views in database queries for the first version of the product. We will only support tables.

When querying, isn't a view the same as a table for all intents? Or, is our view interface reliant on some database structure introspection that won't yet be able to work with other views until after alpha?

**Reply by ghislaineguerin** [reply]
The more I think about the input, output flow, I also feel that creating a "master view" and then consuming those columns in other views would reduce a lot of the effort for users. I can think of different summarisation columns or methods being applied to the same set of columns to produce a set of reports and having to change parameters only in one place.

**Reply by mathemancer** [reply]
@dmos62 It may be even more different at the UI and design level, but the main difference at the SQLAlchemy level (for querying) is reflection, and showing where the data can be edited. If we let users get too far down some chain of views without careful guidance, it may be that they can no longer figure out where to edit if they want to change a student ID number or whatever.

As for the prompt, I think these limitations make sense for the initial alpha.

**Reply by dmos62** [reply]
@mathemancer hm, as I said in my comments in the above sub-thread, I think that being able to build up a view "pipeline", a chain or tree of views, is important, even if we have a stellar query engine.

Maybe we can do some best-effort type of reflection, where we try to point to the previous transformation step (steps, if it's a fork in the pipeline). The user could then navigate the pipeline, one step at a time.

**Reply by kgodey** [reply]
These are limitations for the _alpha_ release, I expect we will support building more complex queries in future releases. I am not sure if this will take the form of allowing Views to be used in other Views or if we will expand our query builder into a pipeline format (effectively allowing people to use auxiliary views while building a final view).

The reasons for not building queries on top of Views for the alpha:
- it's going to be difficult to determine how to join Views to tables used for querying. We are relying on foreign keys as the main method of joining – it also works well with our "Link Table" UX for foreign keys. We've got enough complex problems to solve for the query builder already, I don't want to add this one.
    - If I'm missing something and determining how to join Views will actually be easy, let me know. 
- In order to have auxiliary views to support your main view, the better user experience might actually be to represent the query builder as a pipeline of actions rather than allow users to use random View columns in their queries. But I don't want to start off with a pipeline because it will be complicated to implement.

Does that make sense?

**Reply by dmos62** [reply]
@kgodey I'm aware of the costs involved in switching directions now. As a side note, the reflection/inspection mechanism seems brittle.

**Reply by kgodey** [reply]
> As a side note, the reflection/inspection mechanism seems brittle.

@dmos62 I don't know what you're referring to here, can you elaborate? What is the mechanism you're referring to and how is it brittle?

> I'm aware of the costs involved in switching directions now.

I'm also not sure what you mean here.

**Reply by pavish** [reply]
> Do these limitations and assumptions make sense?

Yes, they do to me.

> Is there anything in this page where the wording is confusing or ambiguous?

Could we mention somewhere/have an alert that says all these decisions are for the _alpha_ and we might re-evaluate them after that? This applies to all views pages.

**Reply by dmos62** [reply]
@kgodey 

> > As a side note, the reflection/inspection mechanism seems brittle.
> 
> @dmos62 I don't know what you're referring to here, can you elaborate? What is the mechanism you're referring to and how is it brittle?
> 
> > I'm aware of the costs involved in switching directions now.
> 
> I'm also not sure what you mean here.

Sorry for the ambiguity. I meant to say that I won't talk about alternatives, because we're already invested in this way of handling views. The reflection mechanism I was talking about is the one that helps users find related data that could be joined into a view. By brittle I meant that it fails or is difficult to implement for some use cases, for example when using with views, which is hence unsupported in alpha. By pointing out that it's brittle, I was hinting that it might be difficult to build upon, and so it would be good to keep it in mind during future refactors.

**Reply by kgodey** [reply]
@dmos62 I think this is a good point to discuss alternatives. The spec is up for discussion before it's finalized so that we can consider alternatives. Even if we don't end up taking an alternative approach, it would be good to know what kinds of use cases the initial version does not handle well. One of the goals for this implementation is to build a foundation that we can hopefully easily extend to handle other use cases in the future.

>  The reflection mechanism I was talking about is the one that helps users find related data that could be joined into a view. 

This is also part of the spec that's under discussion. Do you have a suggestion for an algorithm for how to find related data in other views?

**Reply by dmos62** [reply]
@kgodey I may not have thought about this enough, but how about letting the user pick arbitrary columns to join on and provide best-effort non-authoritative suggestions on top of that? Similarly to what we did with `DBFunction`s and hints. That way it's ok if the suggestions fail to find the data you're interested in (in some cases), because you can resort to inputing the join predicate (the column to join on) manually.

**Comment by kgodey** [comment]
## Feature Requirements

[02. Feature Requirements](https://wiki.mathesar.org/en/product/specs/2022-01-views/02-feature-requirements)

Prompts:
- This page is supposed to describe the features we'll be adding to the product at a high level. Does it meet that goal?

**Reply by dmos62** [reply]
>- formulas to use (translating under the hood to SQL functions)

Nitpick: not sure it's useful to single out formulas when saying that they'll be translated to SQL expressions. Other things on the list will be translated to SQL expressions too. Also, you used the term SQL functions, which might be a bit too specific: for example `DBFunction`s translate to SQL expressions, not functions. Maybe it's too early to go into specific implementation details.

**Reply by mathemancer** [reply]
> * This page is supposed to describe the features we'll be adding to the product at a high level. Does it meet that goal?

I think so.

**Reply by pavish** [reply]
> This page is supposed to describe the features we'll be adding to the product at a high level. Does it meet that goal?

I think it does.

**Comment by kgodey** [comment]
## The Query Builder

[03. The Query Builder](https://wiki.mathesar.org/en/product/specs/2022-01-views/03-the-query-builder)

Prompts:
- Any questions about what features the query builder will and will not support?
- Any questions about a specific concept?
    - Reference table
    - Column selection
    - Query refinement
    - Summarization
    - Previews
- Any questions about implementation details?

**Reply by dmos62** [reply]
An observation about adding a column from a table vs. adding from a formula: both can be reduced to the same process, if you say that adding a column X from a table is adding a column from an "identity" formula parametrized by that column X. "Identity" formula as in an identity function (`lambda x: x`).

**Reply by dmos62** [reply]
> The summarization is applied to the query results as follows:
>
>    - [...]
>    - the values of rest of the columns are aggregated into a list of distinct values.

I can imagine scenarios where I'd want to do something like "bag union" aggregation instead of "set union" (a set union is the equivalent of a list of distinct values in those sets). Also, it's something that depends on the column, so a single summarization ideally would support different aggregations for different columns.

Do we expect that "set union"-only aggregation will cover enough cases to be interesting on its own?

**Reply by ghislaineguerin** [reply]
> This is only applicable to tables which have multiple related records to the reference table.

Having aggregations at the input step would mean that the table is summarised by default. Since we have discussed allowing the denormalised table input and only then applying a summarisation step, I think aggregation would apply only to the summarised output rather than input.

**Reply by mathemancer** [reply]
> Any questions about what features the query builder will and will not support?

Some.  See below.

>   Query refinement

This section makes a lot of assumptions that the main goal will be to aggregate data spread across multiple tables (or at least that's where most of the emphasis of the document lies). I think this is a valid use of querying, but another (very common) is to create "normalized" views (starting as queries) of a denormalized table, both just for viewing and also for consideration when trying to actually normalize a data model. The user needs to be able to look for and assess repetition, not just within a single column, but in tuples composed of multiple columns for that task. For example, if I have a big wide denormalized "roster" table, it would be pretty useful to be able to build a query that shows unique `student_id`, `student_name`, `student_email` tuples. I could either just use that query to (for example) check for mistakes associating one student with multiple IDs.  I could also create a view from that query or even split the results into their own table and start normalizing my DB setup.

**Reply by kgodey** [reply]
> a single summarization ideally would support different aggregations for different columns.

The idea with summarization is that you would choose a single "group by" column and the rest of the columns would be aggregated. We choose a default method of aggregation and you can change it as needed on a column-by-column basis (e.g. you can set the aggregation method to "average" for number columns). I'd like to start off with distinct (set) aggregations so that we have fewer things to convey in the UX. Since we'll have the framework set up for different types of aggregations, we can extend it to bag aggregations in the future if users request it.

> For example, if I have a big wide denormalized "roster" table, it would be pretty useful to be able to build a query that shows unique `student_id`, `student_name`, `student_email` tuples.

Could you provide an example schema and SQL query? I'll need to consider what abstraction to use to represent it in the UX.

**Reply by mathemancer** [reply]
### Scenario

You have a `roster` table for your large university. Suppose the university has 50,000 students, 5,000 professors, and each student takes 5 classes, and each professor teaches around 3 classes. The `roster` table might be:

| student_id | student_name | student_email  | professor_name | subject | grade |
|------------|--------------|----------------|----------------|---------|------:|
| 1234123412 | Alice        | alice@uni.edu  | Bob            | Math    | 82    |
| 1234123412 | Alice        | alice@uni.edu  | Carol          | Reading | 97    |
| 1234123412 | Alice        | alice@uni.edu  | David          | Writing | 92    |
| 1234123412 | Alice        | alice@uno.edu  | Edith          | Science | 72    |
| 1234123412 | Alice        | alice@uni.edu  | Frank          | Music   | 46    |
| 7890789078 | George       | george@uni.edu | Bob            | Math    | 82    |
| 7890789078 | Georg        | george@uni.edu | Henrietta      | Reading | 54    |
| 7890789078 | George       | george@uni.com | Igor           | Writing | 71    |
| 7890789078 | George       | george@uni.edu | Jennifer       | Science | 84    |
| 7890789078 | George       | george@uni.edu | Kenneth        | Music   | 100   |

### Simulated normalization

We want to check for errors, and a normalized schema would help. If we don't actually _want_ a normalized schema (for performance reasons; way we want to avoid joins in our large(ish) table), we might still pluck out what the students table _would_ be with the query:

```sql
SELECT DISTINCT student_id, student_name, student_email
FROM roster
ORDER BY student_id, student_name, student_email;
```

This would result in (using only the subset of data above):

| student_id | student_name | student_email  |
|------------|--------------|----------------|
| 1234123412 | Alice        | alice@uni.edu  |
| 1234123412 | Alice        | alice@uno.edu  |
| 7890789078 | Georg        | george@uni.edu |
| 7890789078 | George       | george@uni.com |
| 7890789078 | George       | george@uni.edu |

Even without actually creating the table and linking with a foreign key, it's easy to see probable mistakes from this view. In cases where you have a denormalized schema, and are therefore subject to data integrity problems, these sorts of queries are super common tasks.

### Analytics

Another common task given the above table would be to get, say, the average grade for each student:

```sql
SELECT student_id, avg(grade) FROM roster GROUP BY student_id;
```

This gives (using our subset above again)

| student_id | avg(grade) |
|------------|-----------:|
| 1234123412 | 77.8       |
| 7890789078 | 78.2       |

I think aggregations of this sort based on groupings of data within a table are super common, even more common than aggregating based on links to other tables.

**Reply by kgodey** [reply]
Thanks for this, it was very helpful.

The average grade for each student is already possible based on the "summarization" feature (which is basically group by).

We don't have a way of selecting multiple distinct columns in the current spec (simulated normalization), but it should be easy to extend the "summarization" feature to do this (currently I have artificially limited to 1 group by, we could instead allow any number of columns). I'll update the spec accordingly.

**Comment by kgodey** [comment]
## Formulas

[04. Formulas](https://wiki.mathesar.org/en/product/specs/2022-01-views/04-formulas). There are also a number of sub-pages defining different types of formulas.

Prompts:
- Do the defined formula variable types make sense?
- Do the variety of formulas defined make sense?
- Does the custom formula type make sense?
- Only the list formula is editable – does that make sense, and does the editing behavior make sense?
- I'm also thinking about putting off regular expression formulas and/or custom formulas until after the alpha – any thoughts on that?
- We may also be able to reuse the formulas for generated columns in tables, any thoughts on that?

**Reply by dmos62** [reply]
I'm trying to imagine how the various elements of this system will come together.

The way we've been using `DBFunction`s so far, we excluded a bunch of uses that *a)* would require to look at more than one row at a time, and *b)* that would affect more than a single column or row.

It seems that a formula is conceptually identical to that, except that it also does *a)*. It can look at more than one row at a time, but it still doesn't decide the value of more than one column (and doesn't change the number of rows).

Maybe this train of thought will be useful.

Edit: in section 07 formulas are reffered to as window functions, which is a better way to put it (than I put it).

**Reply by kgodey** [reply]
I was thinking we'll end up extending `DBFunction`s to implement formulas, and we'll probably have a UI endpoint similar to filters that provide a list of supported formulas to the UI.

**Reply by dmos62** [reply]
@kgodey 

> I was thinking we'll end up extending DBFunctions to implement formulas, and we'll probably have a UI endpoint similar to filters that provide a list of supported formulas to the UI.

That *should* work. I'm not yet sure what the various SQL expressions that we'll need to generate will look like, but essentially `DBFunction` should handle anything. What might need refactoring is the hint system.

**Comment by kgodey** [comment]
## Views

[05. View Structure](https://wiki.mathesar.org/en/product/specs/2022-01-views/05-view-structure)
[06. View Columns](https://wiki.mathesar.org/en/product/specs/2022-01-views/06-view-columns)

Prompts:
- Does View editing behavior make sense?
- Does View column information make sense?

**Reply by ghislaineguerin** [reply]
This looks good. It would be good also if users could see the views in which a particular table is being referenced. Not sure if we'd store that information though.

**Reply by kgodey** [reply]
>  It would be good also if users could see the views in which a particular table is being referenced.

Would this be when the user is looking at the table?

**Reply by silentninja** [reply]
> Does View editing behavior make sense?

Yes, it makes sense

**Comment by kgodey** [comment]
## Extra Information

[07. Breaking Down DB Queries](https://wiki.mathesar.org/en/product/specs/2022-01-views/07-breaking-down-db-queries)
[08. Appendix](https://wiki.mathesar.org/en/product/specs/2022-01-views/08-appendix)

No specific prompts here, but feel free to leave feedback.

**Comment by kgodey** [comment]
## 2022-03-08 Spec Updates

Here's a GitHub diff of the spec updates I made today in response to feedback: https://github.com/centerofci/mathesar-wiki/compare/f8df0e02d6ea67e8af7028a467f3bf3a533c05d4...db399a630aea227f01b362bc40abfe8da38eddbc

- @mathemancer, this should address your concern about the Overview page assumption that users will generally be working with a well normalized database schema.
- @mathemancer @dmos62 I've made the assumption that we're working with joins based on FKs explicit and tied the assumption that we can't use Views in other Views contingent on it. This still seems like the best way to go to keep implementation simple for the alpha release.
- @pavish I've added an infobox in the Assumptions page that says these assumptions are for the alpha release. I don't think the other pages need the same warning because the spec is meant to be defining a set of features we plan to implement at a moment in time; we'll have a new spec for any new features.

I'm still waiting on an example schema from @mathemancer to figure out if we need to create any new concepts to resolve this concern.

> This section makes a lot of assumptions that the main goal will be to aggregate data spread across multiple tables (or at least that's where most of the emphasis of the document lies). I think this is a valid use of querying, but another (very common) is to create "normalized" views (starting as queries) of a denormalized table, both just for viewing and also for consideration when trying to actually normalize a data model. The user needs to be able to look for and assess repetition, not just within a single column, but in tuples composed of multiple columns for that task. For example, if I have a big wide denormalized "roster" table, it would be pretty useful to be able to build a query that shows unique `student_id`, `student_name`, `student_email` tuples. I could either just use that query to (for example) check for mistakes associating one student with multiple IDs. I could also create a view from that query or even split the results into their own table and start normalizing my DB setup.

### Edited to add:
I also added some new variable types and filters in https://github.com/centerofci/mathesar-wiki/compare/db399a630aea227f01b362bc40abfe8da38eddbc...18e58e6111f8327b3c9166a36d8111904517560c

@ghislaineguerin This should address the type of formula we talked about where you can apply a filter to a column and get a boolean result.

**Comment by kgodey** [comment]
## 2022-03-09 Spec Updates

I've updated the requirements for the "summarization" feature in https://github.com/centerofci/mathesar-wiki/commit/016d74e1c54d32dd275585b9554b3e561e874bb8 to include the ability to summarize by multiple columns and add @mathemancer's [example scenario](https://github.com/centerofci/mathesar/discussions/1015#discussioncomment-2324239). This should allow users to find distinct data more easily.

@ghislaineguerin This also affects design requirements, so FYI for you.

### Edited to add:
With that, I'm counting the spec as closed.
